plugins/api/src/main/kotlin/PluginFactoryGenerator.kt

```diff
@@ -0,0 +1,352 @@
+                name = %S,
+                className = %S,
+                description = %S,
+                options = listOf(
+            
+            """.trimIndent(),
+            ortPlugin.name,
+            pluginClass.simpleName.asString(),
+            ortPlugin.description
+        )
+
+        pluginOptions.forEach {
+            add(
+                """
+                |        %T(
+                |            name = %S,
+                |            description = %S,
+                |            type = %T.%L,
+                |            defaultValue = %S,
+                |            isRequired = %L
+                |        ),
+                |
+                """.trimMargin(),
+                PluginOption::class,
+                it.name,
+                it.description,
+                PluginOptionType::class,
+                it.type.name,
+                it.defaultValue,
+                it.isRequired
+            )
+        }
+
+        add(
+            """
+                )
+            )
+            """.trimIndent()
+        )
+    }.build()
+
+    /**
+     * Get the constructor of the plugin class that has a [PluginDescriptor] and a config argument. Throw an
+     * [IllegalArgumentException] if more than one or no such constructor exists.
+     */
+    private fun getPluginConstructor(pluginClass: KSClassDeclaration): KSFunctionDeclaration {
+        // TODO: Consider adding an @OrtPluginConstructor annotation to mark the constructor to use. This could be
+        //       useful if a plugin needs multiple constructors for different purposes like testing.
+        val constructors = pluginClass.getConstructors().filterTo(mutableListOf()) {
+            if (it.parameters.size < 1 || it.parameters.size > 2) {
+                return@filterTo false
+            }
+
+            val firstArgumentIsDescriptor = it.parameters[0].name?.asString() == "descriptor" &&
+                it.parameters[0].type.resolve().declaration.qualifiedName?.asString() ==
+                "org.ossreviewtoolkit.plugins.api.PluginDescriptor"
+
+            val optionalSecondArgumentIsCalledConfig =
+                it.parameters.size == 1 || it.parameters[1].name?.asString() == "config"
+
+            firstArgumentIsDescriptor && optionalSecondArgumentIsCalledConfig
+        }
+
+        require(constructors.size == 1) {
+            "Plugin class $pluginClass must have exactly one constructor with a PluginDescriptor and a config " +
+                "argument."
+        }
+
+        return constructors.first()
+    }
+
+    /**
+     * Get the plugin options from the config class by mapping its properties to [PluginOption] instances.
+     */
+    @OptIn(KspExperimental::class)
+    private fun KSClassDeclaration.getPluginOptions(): List<PluginOption> {
+        require(Modifier.DATA in modifiers) {
+            "Config class $this must be a data class."
+        }
+
+        require(getConstructors().toList().size == 1) {
```

Comment:
There's also `singleOrNull()` for sequences, which could again be used to combine the check with getting the single element.

Reply:
Fixed in a new commit, changing it in place was annoying due to the split of the class in a later commit.